### ISSUE TITLE
Version 2.1.0 does not work on Raspberry Pi (Buster)

### ISSUE THREAD

**Issue by Rak1ta** [00-Evan/shattered-pixel-dungeon#1407]
Upgraded from 2.0.2 to 2.1.0 and encountered a startup error.

```
OS: Raspbian GNU/Linux 10 (buster)
OS architecture: 32-bit
Kernel: armv7l 5.10.103-v7l+
Device model: Raspberry Pi 4 Model B Rev 1.2 
Cpu name: Cortex-A72
Ram size: 3.65 GB
Raspberry Pi OS image version: 2021-05-07
Language: en_US.UTF-8
```

```
pi@raspberrypi:~ $ /usr/lib/jvm/java-11-openjdk-armhf/bin/java -jar /home/pi/.local/bin/ShatteredPD-Desktop.jar
[GAME] com.badlogic.gdx.utils.GdxRuntimeException: Error creating controller manager: com.badlogic.gdx.controllers.desktop.JamepadControllerManager
	at com.badlogic.gdx.controllers.Controllers.initialize(Controllers.java:118)
	at com.badlogic.gdx.controllers.Controllers.addListener(Controllers.java:64)
	at com.watabou.noosa.Game.create(Game.java:106)
	at com.shatteredpixel.shatteredpixeldungeon.ShatteredPixelDungeon.create(ShatteredPixelDungeon.java:70)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.initializeListener(Lwjgl3Window.java:416)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.update(Lwjgl3Window.java:366)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:192)
	at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:166)
	at com.shatteredpixel.shatteredpixeldungeon.desktop.DesktopLauncher.main(DesktopLauncher.java:192)
Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: Couldn't load shared library 'libjamepadarm.so' for target: Linux, ARM, 32-bit
	at com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:128)
	at com.studiohartman.jamepad.ControllerManager.<init>(ControllerManager.java:68)
	at com.studiohartman.jamepad.ControllerManager.<init>(ControllerManager.java:52)
	at com.badlogic.gdx.controllers.desktop.JamepadControllerManager.<init>(JamepadControllerManager.java:32)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at com.badlogic.gdx.utils.reflect.ClassReflection.newInstance(ClassReflection.java:93)
	at com.badlogic.gdx.controllers.Controllers.initialize(Controllers.java:116)
	... 8 more
Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/615224bb/libjamepadarm.so: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/615224bb/libjamepadarm.so)
	at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:325)
	at com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:124)
	... 18 more
Caused by: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/615224bb/libjamepadarm.so: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/615224bb/libjamepadarm.so)
	at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)
	at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2445)
	at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2501)
	at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2700)
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2630)
	at java.base/java.lang.Runtime.load0(Runtime.java:768)
	at java.base/java.lang.System.load(System.java:1837)
	at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:331)
	at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:300)
	... 19 more

[LWJGL] Failed to load a library. Possible solutions:
	a) Add the directory that contains the shared library to -Djava.library.path or -Dorg.lwjgl.librarypath.
	b) Add the JAR that contains the shared library to the classpath.
[LWJGL] Enable debug mode with -Dorg.lwjgl.util.Debug=true for better diagnostics.
[LWJGL] Enable the SharedLibraryLoader debug mode with -Dorg.lwjgl.util.DebugLoader=true for better diagnostics.

Exception: java.lang.UnsatisfiedLinkError thrown from the UncaughtExceptionHandler in thread "main"

```

Is that a mistake?
```
Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: Couldn't load shared library 'libjamepadarm.so' for target: Linux, ARM, 32-bit
Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/615224bb/libjamepadarm.so: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/615224bb/libjamepadarm.so)
Caused by: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/615224bb/libjamepadarm.so: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/615224bb/libjamepadarm.so)
```

**Comment by 00-Evan** [comment]
This looks to be caused by a dependency update in v2.1.0 that causes the game's controller library to no longer work with 32-bit linux (and possibly 32-bit windows and 32-bit macOS).

In general I am not going to try and go out of my way to support 32-bit devices, but this should be a relatively easy fix in v2.1.1. Controller functionality will not work anymore, but the game won't crash on launch ether.

**Comment by theofficialgman** [comment]
> In general I am not going to try and go out of my way to support 32-bit devices, but this should be a relatively easy fix in v2.1.1. Controller functionality will not work anymore, but the game won't crash on launch ether.

doubt it is 32bit specific. the error is clear
```
Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/615224bb/libjamepadarm.so: /lib/arm-linux-gnueabihf/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/615224bb/libjamepadarm.so)
```

who built the binary `libjamepadarm.so`? It needs to be re-built on an older distro (debian stretch or bionic) so that the GLIBC version is lowered to something like 2.17 for compatability.

**Comment by theofficialgman** [comment]
found upstream, you should report the issue there https://github.com/libgdx/Jamepad as its a dependency of https://github.com/libgdx/gdx-controllers . @MrStahlfelge runs both repos. Can you explain what system you used to build these native binaries it? They aren't widely compatible and I suggest either installing an older distro like ubuntu bionic/debian stretch to build on, or using a chroot with those versions.

**Comment by 00-Evan** [comment]
Apologies, I saw 32-bit and assumed it was relating to that, as I've seen issues on 32-bit systems before.

Would be nice to get this fixed at the source, but I have at least taken steps to mitigate this and future crashes of the sort in Shattered. Controller support isn't necessary for the game to work, obviously, so in v2.1.1 if initialization fails the game will catch the error and just turn controller support off.

**Comment by MrStahlfelge** [comment]
> Can you explain what system you used to build these native binaries it? 

GitHub actions, see the commit history for relevant changes and feel free to submit PRs changing what you need.

**Comment by theofficialgman** [comment]
@MrStahlfelge I plan on it. got ARM64/ARMhf working with minimum GLIBC 2.17 already. https://github.com/theofficialgman/Jamepad . Having issues with x86 and x86_64 currently.

**Comment by MrStahlfelge** [comment]
The build should run on GitHub actions, otherwise I would suggest to make an own fork.

**Comment by theofficialgman** [comment]
Yup these are actions changes, guess you didn't actually look at it https://github.com/theofficialgman/Jamepad/blob/master/.github/workflows/pushaction.yml  Patches to libc just like chromium does for their sysroots so that the resultant builds link to older GLIBC versions.

**Comment by theofficialgman** [comment]
PR is complete. this issue can be closed if @MrStahlfelge merges it and creates a new release and then @00-Evan publishes a new version using it.

**Comment by 00-Evan** [comment]
I plan to close this after releasing v2.1.1, which includes a mitigation for the crash, but obviously I'll be happy to also update gdx-controllers when that's available.

**Comment by MrStahlfelge** [comment]
Probably using an older version (compiled on an older ubuntu version) would work as well.

**Comment by 00-Evan** [comment]
fixed in commit e9ccd7929bb43bdf22d15e00c0d9a422abfe5fb3

**Comment by theofficialgman** [comment]
@hyppyp please post the terminal log when you try to launch here showing that 2.1.1 crashes on launch

**Comment by hyppyp** [comment]
Sorry for being stupid, where do I find the terminal log?

 

 

Van: theofficialgman ***@***.*** 
Verzonden: woensdag 14 juni 2023 17:23
Aan: 00-Evan/shattered-pixel-dungeon
CC: hyppyp; Mention
Onderwerp: Re: [00-Evan/shattered-pixel-dungeon] Version 2.1.0 does not work on Raspberry Pi (Buster) (Issue #1407)

 

@hyppyp <https://github.com/hyppyp>  please post the terminal log when you try to launch here showing that 2.1.1 crashes on launch

**Comment by theofficialgman** [comment]
You just run it via terminal according to the pi-apps description for the app.
```
/usr/lib/jvm/java-11-openjdk-arm64/bin/java -jar -jar $HOME/.local/bin/ShatteredPD-Desktop.jar
```

**Comment by hyppyp** [comment]
$ /usr/lib/jvm/java-11-openjdk-arm64/bin/java -jar $HOME/.local/bin/ShatteredPD-Desktop.jar

[ALSOFT] (EE) Failed to set real-time priority for thread: Actie is niet toegestaan (1)

[ALSOFT] (EE) Failed to set real-time priority for thread: Actie is niet toegestaan (1)

[GAME] com.badlogic.gdx.utils.GdxRuntimeException: Error creating controller manager: com.badlogic.gdx.controllers.desktop.JamepadControllerManager

            at com.badlogic.gdx.controllers.Controllers.initialize(Controllers.java:118)

            at com.badlogic.gdx.controllers.Controllers.addListener(Controllers.java:64)

            at com.watabou.noosa.Game.create(Game.java:106)

            at com.shatteredpixel.shatteredpixeldungeon.ShatteredPixelDungeon.create(ShatteredPixelDungeon.java:69)

            at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.initializeListener(Lwjgl3Window.java:416)

            at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Window.update(Lwjgl3Window.java:366)

            at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.loop(Lwjgl3Application.java:192)

            at com.badlogic.gdx.backends.lwjgl3.Lwjgl3Application.<init>(Lwjgl3Application.java:166)

            at com.shatteredpixel.shatteredpixeldungeon.desktop.DesktopLauncher.main(DesktopLauncher.java:188)

Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: Couldn't load shared library 'libjamepadarm64.so' for target: Linux, ARM, 64-bit

            at com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:128)

            at com.studiohartman.jamepad.ControllerManager.<init>(ControllerManager.java:68)

            at com.studiohartman.jamepad.ControllerManager.<init>(ControllerManager.java:52)

            at com.badlogic.gdx.controllers.desktop.JamepadControllerManager.<init>(JamepadControllerManager.java:32)

            at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)

            at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)

            at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)

            at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)

            at java.base/java.lang.Class.newInstance(Class.java:584)

            at com.badlogic.gdx.utils.reflect.ClassReflection.newInstance(ClassReflection.java:93)

            at com.badlogic.gdx.controllers.Controllers.initialize(Controllers.java:116)

            ... 8 more

Caused by: com.badlogic.gdx.utils.SharedLibraryLoadRuntimeException: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/75a7d97b/libjamepadarm64.so: /lib/aarch64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/75a7d97b/libjamepadarm64.so)

            at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:325)

            at com.badlogic.gdx.utils.SharedLibraryLoader.load(SharedLibraryLoader.java:124)

            ... 18 more

Caused by: java.lang.UnsatisfiedLinkError: /tmp/libgdxpi/75a7d97b/libjamepadarm64.so: /lib/aarch64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by /tmp/libgdxpi/75a7d97b/libjamepadarm64.so)

            at java.base/java.lang.ClassLoader$NativeLibrary.load0(Native Method)

            at java.base/java.lang.ClassLoader$NativeLibrary.load(ClassLoader.java:2445)

            at java.base/java.lang.ClassLoader$NativeLibrary.loadLibrary(ClassLoader.java:2501)

            at java.base/java.lang.ClassLoader.loadLibrary0(ClassLoader.java:2700)

            at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2630)

            at java.base/java.lang.Runtime.load0(Runtime.java:768)

            at java.base/java.lang.System.load(System.java:1837)

            at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:331)

            at com.badlogic.gdx.utils.SharedLibraryLoader.loadFile(SharedLibraryLoader.java:300)

            ... 19 more

 

***@***.***:~ $

 

Van: theofficialgman ***@***.*** 
Verzonden: woensdag 14 juni 2023 18:00
Aan: 00-Evan/shattered-pixel-dungeon
CC: hyppyp; Mention
Onderwerp: Re: [00-Evan/shattered-pixel-dungeon] Version 2.1.0 does not work on Raspberry Pi (Buster) (Issue #1407)

 

You just run it via terminal according to the pi-apps description for the app.

/usr/lib/jvm/java-11-openjdk-arm64/bin/java -jar $HOME/.local/bin/ShatteredPD-Desktop.jar

**Comment by 00-Evan** [comment]
Yeah I'm dumb and missed a 'return false'. Will properly correctly in v2.1.2 shortly

**Comment by 00-Evan** [comment]
fixed (properly this time) in commit aede0d6a484cb527e826952c32e36b91077f06df
